### PR TITLE
Add Additional Iterator Functions

### LINKED ISSUE #2028: Iterator
Add the following functions to iterator:

Phase 1:

- [ ] `head`
- [ ] `range`
- [ ] `repeat`
- [ ] `memberOf`
- [ ] `toArray`
- [ ] `toMap`
- [ ] `toSet`

Phase 2:

- [ ] `replace`
- [ ] `exists`
- [ ] `forall`
- [ ] `dropWhile`
- [ ] `foreach`
- [ ] `zip`
- [ ] `takeWhile` - Eager/Lazy split
- [ ] `zipWith` - Eager/Lazy split
- [ ] `foldLeft`
- [ ] `foldRight`
- [ ] `findRight`?

Phase 3:

- [ ] `append`
- [ ] `filterMap` - Eager/Lazy split
- [ ] `findMap` - Eager/Lazy split
- [ ] `reduceLeft`
- [ ] `reduceRight`
- [ ] `intersperse`
- [ ] `minimum`
- [ ] `minimumBy`
- [ ] `maximum`
- [ ] `maximumBy`

Phase 4:

- [ ] `mapWithIndex`
- [ ] `flatMap`
- [ ] `intercalate`
- [ ] `flatten`
- [ ] `partition`
- [ ] `span`

See `List.flix` for reference

### PR DESCRIPTION
**Pull request by jaschdoc**

### CODE REVIEW

File: main/src/library/Iterator.flix [hunk 1]
```diff
@@ -20,6 +20,60 @@ pub enum Iterator[a] {
...
+    /// Returns `None` if `iter` is empty.
+    ///
+    /// Consumes the first element of the iterator.
+    ///
+    pub def head(iter: Iterator[a]): Option[a] & Impure = ???
+
+    ///
+    /// Returns an iterator of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns `Done` if `b >= e`.
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:35]: Returns an empty iterator if ...

File: main/src/library/Iterator.flix [hunk 2]
```diff
@@ -20,6 +20,60 @@ pub enum Iterator[a] {
...
+    ///
+    pub def range(b: Int32, e: Int32): Iterator[Int32] & Impure = ???
+
+    ///
+    /// Returns an iterator over an iterable with the element `a` repeated `n` times.
+    ///
+    /// Returns `Done` if `n < 0`.
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:42]: Same as above

File: main/src/library/Iterator.flix [hunk 3]
```diff
@@ -20,6 +20,60 @@ pub enum Iterator[a] {
...
+    ///
+    pub def repeat(a: a, n: Int32): Iterator[a] & Impure = ???
+
+    ///
+    /// Returns `true` if and only if `iter` contains the element `a`.
+    ///
+    // TODO: Does this consume elements until next() == a?
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:49]: Yes, I would say:

Consumes elements of the iterator until an element equal to `a` is found.

File: main/src/library/Iterator.flix [hunk 4]
```diff
@@ -20,6 +20,66 @@ pub enum Iterator[a] {
...
+        else
+            Some(next())
+    }
+
+    ///
+    /// Returns an iterator of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns an empty iterator if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Iterator[Int32] & Impure = List.range(b, e) |> toIter
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:43]: This constructs the entire list and then makes it lazy. That works, but is not so great if the range is longer.

Instead, you should use a counter and construct two functions `done` and `next`. (See the documentation on how to use references).

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 5]
```diff
@@ -1,5 +1,57 @@
...
+    @test
+    def range04(): Bool & Impure =
+        let iter = Iterator.range(-1, 2);
+        let minusOne = Iterator.head(iter);
+        let zero = Iterator.head(iter);
+        let one = Iterator.head(iter);
+        Some(-1) == minusOne and
+        Some(0) == zero and
+        Some(1) == one and
+        None == Iterator.head(iter)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:48]: I think these tests are too complicated.

It is okay to just to `Iter.range(1, 1) |> Iter.toList == theListYouExpect`

File: main/src/library/Iterator.flix (hunk 4 above)
Comment (jaschdoc) [reply main/src/library/Iterator.flix:43]: I see, thanks.
My intention was to make a "dumb" implementation first, make the tests pass and then redo it.

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix (hunk 5 above)
Comment (jaschdoc) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix]: Fixed it! :)

**Comment by jaschdoc** [comment]
I feel like I'm missing some test cases for `toMap` and `toSet`, maybe you have some feedback on that?

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 6]
```diff
@@ -391,6 +391,74 @@ namespace TestIterator {
...
+        let list = List.zip(List.range(0, 500), List.repeat("b", 500));
+        zipped |> Iterator.toList == list and
+        zipped |> Iterator.isEmpty and
+        Iterator.zip(iterB, iterA) |> Iterator.isEmpty
+
+    @test
+    def zip08(): Bool & Impure =
+        let iter = Iterator.range(0, 500);
+        let list = List.range(0, 500);
+        Iterator.zip(iter, iter) |> Iterator.toList == List.zip(list, list) and
```
Comment (jaschdoc) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:458]: Should zipping an iterator with itself produce a tuple `(x, x)` or `(x, y)`, where `x` is the element at position `i` in an iterator and `y` the element at position `i + 1`?
Intuitively, it makes sense that it would produce `(x, x)`

**Comment by magnus-madsen** [comment]
This PR is getting big. I suggest that I do a round of feedback and then you fix those and we do a merge.

Then we can open a new PR for the next round of improvements.

File: main/src/library/Iterator.flix [hunk 7]
```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
...
+    ///
+    pub def memberOf(a: a, iter: Iterator[a]): Bool & Impure with Eq[a] =
+        let Iterator(done, next) = iter;
+        if (done())
+            false
+        else
+            next() == a or memberOf(a, iter)
+
+    ///
+    /// Returns the iterator `iter` as an array.
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:80]: I would rephrase: `Returns all elements in the iterator iter as an array`.

(Update similar comments.)

File: main/src/library/Iterator.flix [hunk 8]
```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
...
+            f(next()) and forall(f, iter)
+
+    ///
+    /// Returns `iter` without the longest prefix that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    ///
+    pub def dropWhile(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure =
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:192]: Add a TODO that this likely needs to be overloaded based on the purity/impurity of f.

File: main/src/library/Iterator.flix [hunk 9]
```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
...
+            if (f(x)) dropWhile(f, iter) else Iterator(done, n)
+
+    ///
+    /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    ///
+    pub def takeWhile(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure = ???
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:214]: Same TODO as above

File: main/src/library/Iterator.flix [hunk 10]
```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
...
+    ///
+    /// Consumes elements of both `iterA` and `iterB` equal to the length of the shortest iterator.
+    /// If `iterA` and `iterB` are equal length, both iterators are fully consumed.
+    ///
+    pub def zip(iterA: Iterator[a], iterB: Iterator[b]): Iterator[(a, b)] =
+        let Iterator(doneA, nextA) = iterA;
+        let Iterator(doneB, nextB) = iterB;
+        let doneAB = () -> doneA() or doneB();
+        let nextAB = () -> (nextA(), nextB());
+        Iterator(doneAB, nextAB)
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:245]: Looks good!

File: main/src/library/Iterator.flix [hunk 11]
```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
...
+    ///
+    /// Returns an iterator where the element at index `i` is `f(a, b)` where
+    /// `a` is the element at index `i` in `iterA` and `b` is the element at index `i` in `iterB`.
+    ///
+    /// If either `iterA` or `iterB` becomes depleted, then no further elements are added to the resulting list.
+    ///
+    /// Consumes elements of both `iterA` and `iterB` equal to the length of the shortest iterator.
+    /// If `iterA` and `iterB` are equal length, both iterators are fully consumed.
+    ///
+    pub def zipWith(f: (a, b) -> c & e, iterA: Iterator[a], iterB: Iterator[b]): Iterator[c] & Impure = ???
```
Comment (magnus-madsen) [inline main/src/library/Iterator.flix:257]: Add a todo similar to the above

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 12]
```diff
@@ -1,8 +1,468 @@
...
+    def range01(): Bool & Impure =
+        Iterator.range(0, 0) |> Iterator.isEmpty
+
+    @test
+    def range02(): Bool & Impure =
+        Iterator.range(10, 0) |> Iterator.isEmpty
+
+    @test
+    def range03(): Bool & Impure =
+        (Iterator.range(1, 4) |> Iterator.toList) == (1 :: 2 :: 3 :: Nil)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:30]: Good, these are the kind of simple tests that I like!

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 13]
```diff
@@ -1,8 +1,468 @@
...
+    @test
+    def memberOf03(): Bool & Impure =
+        let el = "A";
+        let iter = Iterator.repeat(el, 5);
+        iter |> Iterator.memberOf(el) and
+        iter |> Iterator.memberOf(el) and
+        iter |> Iterator.memberOf(el) and
+        iter |> Iterator.memberOf(el) and
+        iter |> Iterator.memberOf(el) and
+        iter |> Iterator.isEmpty
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:85]: I think this test is too complicated. How about just constructing some ranges and asking for elements that are there or are not there.

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 14]
```diff
@@ -1,8 +1,468 @@
...
+
+    @test
+    def toArray01(): Bool & Impure =
+        (Nil: List[Int32] |> Iterator.toIter |> Iterator.toArray) == []
+
+    @test
+    def toArray02(): Bool & Impure =
+        let iter = Iterator.range(-10, -5);
+        (iter |> Iterator.toArray) == [-10, -9, -8, -7, -6] and
+        iter |> Iterator.isEmpty
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:107]: Good, but I don't think we should check that the iterator is empty. That is is the right array should be enough.

Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix:107]: I know in principle it is possible to implement it incorrect where it returns an array and still leaves elements in the iterator, but that seems unlikely, and I prefer simpler tests.

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix [hunk 15]
```diff
@@ -1,8 +1,468 @@
...
+
+    @test
+    def replace01(): Bool & Impure =
+        Nil: List[Int32] |> Iterator.toIter |> Iterator.toList ==
+        Nil: List[Int32] |> Iterator.toIter |> Iterator.replace(1, 2) |> Iterator.toList
+
+    @test
+    def replace02(): Bool & Impure =
+        Iterator.repeat("a", 10) |> Iterator.replace("a", "b") |> Iterator.toList ==
+        Iterator.repeat("b", 10) |> Iterator.toList
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestIterator.flix:180]: Could just have the expected list immediately, e.g. `List.repeat(...)`

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix (hunk 6 above)
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix:458]: Zipping an iterator with itself should produce elements (x, x+1). At least in my mind.

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix (hunk 13 above)
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix:85]: That is, for the "expected value", let us just write the list. Either explicitly `1 :: 2 :: Nil` or using something from `List`.

File: main/test/ca/uwaterloo/flix/library/TestIterator.flix (hunk 14 above)
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix:107]: OK, I see you have invested a lot of effort into this, so lets instead keep these tests as they are.

Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestIterator.flix:107]: And importantly you have been very consistent, so thats good!

**Comment by magnus-madsen** [comment]
I think this is very nice work and good progress! I suggest to fix the few issues pointed out and then we can do a merge.

**Comment by magnus-madsen** [comment]
Great! Ready to move on to the next batch of functions :)
